vehicle/kia.go

```diff
@@ -0,0 +1,486 @@
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, uri, bytes.NewReader(dataj))
+	if err != nil {
+		return err
+	}
+	for k, v := range map[string]string{
+		"Authorization":       v.accToken,
+		"ccsp-device-id":      v.deviceID,
+		"ccsp-application-id": "693a33fa-c117-43f2-ae3b-61a02d24f417",
+		"offset":              "1",
+		"Content-Type":        "application/json;charset=UTF-8",
+		"Content-Length":      "72",
+		"Host":                "prd.eu-ccapi.kia.com:8080",
+		"Connection":          "close",
+		"Accept-Encoding":     "gzip, deflate",
+		"User-Agent":          "okhttp/3.10.0",
+	} {
+		req.Header.Set(k, v)
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	return nil
+}
+
+func (v *Kia) sendPIN() error {
+	data := map[string]interface{}{
+		"deviceId": v.deviceID,
+		"pin":      string(v.pin),
+	}
+
+	dataj, err := json.Marshal(data)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPut, kiaUrlSendPIN, bytes.NewReader(dataj))
+	if err != nil {
+		return err
+	}
+	for k, v := range map[string]string{
+		"Authorization":   v.accToken,
+		"Content-type":    "application/json;charset=UTF-8",
+		"Content-Length":  "64",
+		"Host":            "prd.eu-ccapi.kia.com:8080",
+		"Connection":      "close",
+		"Accept-Encoding": "gzip, deflate",
+		"User-Agent":      "okhttp/3.10.0",
+	} {
+		req.Header.Set(k, v)
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	body, _ := ioutil.ReadAll(resp.Body)
+
+	var bbody map[string]interface{}
+	err = json.Unmarshal([]byte(body), &bbody)
+	if err != nil {
+		return err
+	}
+
+	v.controlToken = "Bearer " + fmt.Sprint(bbody["controlToken"])
+	fmt.Println("Control Token: ", v.controlToken)
+
+	return nil
+}
+
+func (v *Kia) getStatus() (float64, error) {
+	uri := kiaUrlGetStatus + v.vehicleID + "/status"
+	req, err := http.NewRequest(http.MethodGet, uri, nil)
+	if err != nil {
+		return 0.0, err
+	}
+	for k, v := range map[string]string{
+		"Authorization":  v.controlToken,
+		"ccsp-device-id": v.deviceID,
+		"Content-Type":   "application/json",
+	} {
+		req.Header.Set(k, v)
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return 0.0, err
+	}
+	defer resp.Body.Close()
+
+	body, _ := ioutil.ReadAll(resp.Body)
+
+	var kr kiaBatteryResponse
+	err = json.Unmarshal([]byte(body), &kr)
+	if err != nil {
+		return 0, err
+	}
+	stateOfCharge := kr.ResMsg.EvStatus.BatteryStatus
+
+	fmt.Println("SoC: ", stateOfCharge)
+
+	return stateOfCharge, nil
+}
+
+// now we have all the needed functions to read Kia SoC
+// chargeState implements the Vehicle.ChargeState interface
+func (v *Kia) chargeState() (float64, error) {
+	fmt.Println("SoC Abfrage gestartet")
+	err := v.getDeviceID()
+	if err != nil {
+		return 0, errors.New("could not obtain deviceID")
+	}
+
+	err = v.getCookies()
```

Comment:
sowas kannst du als einzeiler schreiben:

    if err := ...; err !=nil {

Reply:
alternativ jeweils den nächsten Schritt nur machen falls err==nil, also die Abfrage umdrehen- das erspart die ganzen zusätzlichen Returns

Reply:
ok; formatiert mir vscode beim Speichern aber wieder in 3 Zeilen :-(

Reply:
Yep, aber du sparst jeweils 1 zeile fürs return ein